sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_outlier_filter/cuda_polar_voxel_outlier_filter.cu

```diff
@@ -0,0 +1,949 @@
+      generate_field_data_composer<::cuda::std::optional<int32_t>>(num_points, stream_, mem_pool_);
+
+    FieldDataComposer<double> resolutions{};
+    resolutions.radius = params.radial_resolution_m;
+    resolutions.azimuth = params.azimuth_resolution_rad;
+    resolutions.elevation = params.elevation_resolution_rad;
+
+    dim3 block_dim(512);
+    dim3 grid_dim((num_points + block_dim.x - 1) / block_dim.x, 3, 1);
+
+    switch (polar_type) {
+      case PolarDataType::PreComputed:
+        polar_to_polar_voxel_kernel<float><<<grid_dim, block_dim, 0, stream_>>>(
+          input_cloud->data.get(), num_points, input_cloud->point_step, offsets, resolutions,
+          params.min_radius_m, params.max_radius_m, polar_voxel_indices);
+        break;
+      case PolarDataType::DeriveFromCartesian:
+        cartesian_to_polar_voxel_kernel<float, float><<<grid_dim, block_dim, 0, stream_>>>(
+          input_cloud->data.get(), num_points, input_cloud->point_step, offsets, resolutions,
+          params.min_radius_m, params.max_radius_m, polar_voxel_indices);
+        break;
+      default:
+        throw std::runtime_error("undefined polar_data_type is specified");
+    }
+
+    // calculate unique voxel index that each point belongs to
+    std::tie(num_total_voxels, point_indices, voxel_indices) =
+      calculate_voxel_index(polar_voxel_indices, num_points);
+
+    //  count the number of primary/secondary returns of each voxel
+    num_primary_returns =
+      autoware::cuda_utils::make_unique<size_t>(num_total_voxels, stream_, mem_pool_);
+    num_secondary_returns =
+      autoware::cuda_utils::make_unique<size_t>(num_total_voxels, stream_, mem_pool_);
+    is_in_visibility_range =
+      autoware::cuda_utils::make_unique<int>(num_total_voxels, stream_, mem_pool_);
+    is_primary_returns = autoware::cuda_utils::make_unique<bool>(num_points, stream_, mem_pool_);
+    is_secondary_returns = autoware::cuda_utils::make_unique<bool>(num_points, stream_, mem_pool_);
+
+    // Add points to appropriate vector based on return type
+    size_t return_type_offset = get_offset(input_cloud->fields, "return_type");
+    size_t intensity_offset = get_offset(input_cloud->fields, "intensity");
+    grid_dim = dim3((num_points + block_dim.x - 1) / block_dim.x);
+    classify_point_by_return_type_and_intensity_kernel<uint8_t, uint8_t>
+      <<<grid_dim, block_dim, 0, stream_>>>(
+        input_cloud->data.get(), num_points, num_total_voxels, input_cloud->point_step,
+        return_type_offset, intensity_offset, primary_return_type_dev_.value(),
+        static_cast<uint8_t>(params.intensity_threshold), point_indices.get(), voxel_indices.get(),
+        radius_idx.get(), resolutions.radius, params.visibility_estimation_max_range_m,
+        num_primary_returns.get(), num_secondary_returns.get(), is_in_visibility_range.get(),
+        is_primary_returns.get(), is_secondary_returns.get());
+  }
```

Comment:
That's a lot of arguments for a single function. Probably breaking it down into smaller functions would help.